### ISSUE TITLE
TabsController - unexpected appearance transitions

### ISSUE THREAD

**Issue by markst** [CosmicMind/Material#919]
I'm struggling to figure if this is a iOS 11 bug or Motion bug. 

I have a `TabsController` setup with three view controllers & each of the view controllers view lifecycle functions are called even when the view is not visible.

`viewDidAppear()` is being called despite being a page tab which not visible.

**Comment by danieldahan** [comment]
This is an issue that I am working on for the TabsController. It happens because we need to load the view controllers and get the `tabItem`. I need to fix it, so that I can extract the `tabItem` and not actually need to add the view to the hierarchy. Thoughts?

**Comment by markst** [comment]
Interesting. but there must be some transition calls called on the view controller? it should be possible to load a view controllers view without performing it's view lifecycle events.

@danieldahan

**Comment by danieldahan** [comment]
Yes, agreed. I haven't had a chance to look into it. I will put it as a high-priority as many people use the `TabsController`.

**Comment by markst** [comment]
Just had a quick crack at my setup with `UIPageViewController` and can confirm that calling:
`self.availableViewController.loadViewIfNeeded()`
did not trigger any of the view lifecycle events until that page actually was about to become visible.

**Comment by danieldahan** [comment]
Nice, okay, I will look into it with that approach. Unless you have a moment to dive in. I am working on another issue at the moment.

**Comment by markst** [comment]
@danieldahan nah I wouldn't rush into switching to `UIPageViewController`. sure we can figure a solution. will look into it quick - but afraid I've also got more pressing issues.

**Comment by danieldahan** [comment]
Yeah, I have no plan on using the `UIPageViewController`, as I had originally made the `TabsController` with it and it did not offer much. I think there is a way to ask the view to load but not for rendering, which will probably be an approach to look into.

**Comment by danieldahan** [comment]
As mentioned, it is to get the `tabItem` value from the controller. I need to work on an approach that get's that value without needing to go through the steps you described.

**Comment by markst** [comment]
@danieldahan not sure how `tabItem` has anything to do with what I just subscribed. 

Knock out: `prepareViewController(at: i)` and the problem this issue is describing goes away.

```swift
func prepareViewControllers() {
    for i in 0..<viewControllers.count {
        guard i != selectedIndex else {
            continue
        }
        
        viewControllers[i].view.isHidden = true
        /*
        prepareViewController(at: i)
         */
    }
    
    prepareViewController(at: selectedIndex)
    prepareRootViewController()
}
```

**Comment by danieldahan** [comment]
In order to get the `tabItem`, the view needs to be loaded. It is the only reason I am loading the view controllers in the first place. If it is working without needing this step, then great :)

**Comment by markst** [comment]
@danieldahan that may well be the case. but it's not the reason for this issue I'm describing. 

`TabsController.prepareViewControllers()` is triggering the adding of all the tabs view controller views & adding them to the container view - which doesn't take full advantage of dynamically adding the view controllers view's to the container as and when needed.

**Comment by danieldahan** [comment]
Yes :) I know it is triggering it, and I am doing it for the reason I described earlier. As mentioned, I know the issue and I know the solution I would be looking for, I haven't had a chance to work on it yet.

**Comment by danieldahan** [comment]
Excellent, thank you for the PR. I will review it and push a release :)

**Comment by danieldahan** [comment]
So the PR that you sent doesn't work. It causes the issue that I was describing. Are your TabItems showing up?

**Comment by markst** [comment]
@danieldahan strange. mine works as I expect & tab items are showing up

**Comment by danieldahan** [comment]
Can you try your version with the [TabsController](https://github.com/CosmicMind/Samples/tree/master/Projects/Programmatic/TabsController) sample project? It is not working for me, or maybe share a sample that works? I really want to take your PR, so maybe the setup needs to be updated in the sample project.

**Comment by markst** [comment]
@danieldahan this is simply because your demo project tab item titles are setup on `viewDidLoad()`. 

There is two solutions around that - either you setup the titles on init rather than relying on the controllers view to be loaded. Or have all your views loaded at the time of prepare view controllers: `TabsController.prepareViewControllers()` such as:

```swift
fileprivate extension TabsController {
    /// Prepares all the view controllers.
    func prepareViewControllers() {
        for v in self.viewControllers {
            if #available(iOS 9.0, *) {
                v.loadViewIfNeeded()
            } else {
                _ = v.view
            }
        }
        prepareViewController(at: selectedIndex)
        prepareRootViewController()
    }
```

The disadvantage of forcing the views to load is this will be much larger impact on memory and processing power if there are lots of view controllers as part of a `TabsController`. With the changes on my PR - view controller views will only be loaded as and when necessary.

**Comment by danieldahan** [comment]
Nice, I was hoping to get your PR in. Would you please submit the PR again? Thank you!

**Comment by markst** [comment]
@danieldahan well I don't want to introduce any breaking changes. Perhaps I should include the above logic which forces the view load to? - with a comment mentioning why?

**Comment by danieldahan** [comment]
This has been released in [Material 2.12.4](https://github.com/CosmicMind/Material/releases/tag/2.12.4). Thank you and all the best!
